diff --git a/config.sub b/config.sub
index 7ffe373..f1a9812 100755
--- a/config.sub
+++ b/config.sub
@@ -364,8 +364,6 @@ case $basic_machine in
 	  ;;
 	# Object if more than one company name word.
 	*-*-*)
-		echo Invalid configuration \`$1\': machine \`$basic_machine\' not recognized 1>&2
-		exit 1
 		;;
 	# Recognize the basic CPU types with company name.
 	580-* \
@@ -1321,8 +1319,6 @@ case $basic_machine in
 		# Make sure to match an already-canonicalized machine name.
 		;;
 	*)
-		echo Invalid configuration \`$1\': machine \`$basic_machine\' not recognized 1>&2
-		exit 1
 		;;
 esac
 
